# 01. Import libraries

In [2]:
import pandas as pd
import numpy as np
import os

# 02. Data files

In [3]:
working_directory = os.getcwd()
print(working_directory)

/Users/schitchards/Documents/Data Analytics/Programming/Instacart Basket Analysis/03 Scripts


In [4]:
#create path to data folder
path = r'/Users/schitchards/Documents/Data Analytics/Programming/Instacart Basket Analysis'

In [5]:
#join path to import products.csv
df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col = False)

In [6]:
#join path to import orders_wranged.csv
df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled.csv'), index_col = False)

# 03. test dataframe

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

In [8]:
# Create a mixed type column
df_test['mix'] = ['a','b', 1, True]

In [9]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Checking for mixed-type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)


mix


In [11]:
# Convert columns data type to string
df_test['mix'] = df_test['mix'].astype('str')

In [12]:
# Finding missing values count
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
# Create a subset of the dataframe containing only the missing values in question
df_nan = df_prods[df_prods['product_name'].isnull() ==True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
# Find the mean
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [16]:
# Find the median
df_prods.median()

/var/folders/73/tyj5z5hd1dldnc7qs0ncc1_c0000gn/T/ipykernel_29805/2643079442.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  df_prods.median()


product_id       24845.0
aisle_id            69.0
department_id       13.0
prices               7.1
dtype: float64

In [17]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [18]:
# Compare current dataframe rows with the number in your subset once missing rows have been removed
df_prods.shape

(49693, 5)

In [19]:
# Create new dataframe with non-missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() ==False]

In [20]:
# Run shape again to check that the number of rows has decreased
df_prods_clean.shape

(49677, 5)

In [21]:
# Second way to drop missing values from a particular column
df_prods.dropna(subset = 'product_name', inplace = True)

In [22]:
#Look for duplicate with dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [23]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [24]:
# Addressing Duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [25]:
# Create new dataframe that doesn't include duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [26]:
# Check against original row count to make sure duplicates were removed
df_prods_clean_no_dups.shape

(49672, 5)

In [27]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'))

# Task 4.5

In [28]:
# 2. Consistency checks on df_prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49677.000000  49677.000000   49677.000000  49677.000000
mean   24850.194235     67.763110      11.728687      9.993164
std    14340.588602     38.316396       5.850651    453.592708
min        1.000000      1.000000       1.000000      1.000000
25%    12433.000000     35.000000       7.000000      4.100000
50%    24851.000000     69.000000      13.000000      7.100000
75%    37267.000000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

The max price should be further looked into to see if its an outlier or a mistake.

In [29]:
#3. Check for mixed-type data
for col in df_prods.columns.tolist():
  weird = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[weird]) > 0:
    print (col)

In [30]:
# 4. Change 'product_name' data-type to string
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [31]:
df_prods['product_name'] = df_prods['product_name'].astype('str')

In [32]:
print(pd.api.types.infer_dtype(df_prods['product_name']))

string


In [33]:
#5 Check for missing values in df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
transaction_count              0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

days since prior order column has 206209 missing values.We can see below that there is already a column for 0 days so we know that this does not represent people who have purchased within the past day. The values could represent customers who signed up but have not yet purchased. 

In [34]:
df_ords['days_since_prior_order'].value_counts(dropna=False)

30.0    369323
7.0     320608
6.0     240013
4.0     221696
3.0     217005
5.0     214503
NaN     206209
2.0     193206
8.0     181717
1.0     145247
9.0     118188
14.0    100230
10.0     95186
13.0     83214
11.0     80970
12.0     76146
0.0      67755
15.0     66579
16.0     46941
21.0     45470
17.0     39245
20.0     38527
18.0     35881
19.0     34384
22.0     32012
28.0     26777
23.0     23885
27.0     22013
24.0     20712
25.0     19234
29.0     19191
26.0     19016
Name: days_since_prior_order, dtype: int64

In [35]:
#6 Address the missing values in days_since_prior_order using flag
df_ords.shape

(3421083, 7)

I believe the best approach to addressing the missing values would be to flag them as 'innactive' this way if they become active again their customer data remains in the dataframe.

In [36]:
df_ords_clean = df_ords
df_ords_clean['active'] = df_ords['days_since_prior_order'].isnull() == False

In [37]:
#7 Check for duplicate values in df_ords_clean
df_dups=df_ords_clean[df_ords_clean.duplicated()]

There are no duplicates in the df_ords_clean dataframe

In [38]:
#8 Address any duplicates

There are no duplicates. If there were I would check the shape of the df_ords_clean dataframe (df_ords_clean.shape) then create new dataframe (df_ords_clean_no_dups = df_ords_clean.drop_duplicates()). Finally I would compare the shape of the new dataframe against the old to make sure all duplicates are filtered out.

In [39]:
df_ords_clean.to_csv(os.path.join(path,'02 Data','Prepared Data','orders_checked.csv'))